File: scripts/extract_unseen.mjs
#!/usr/bin/env node
/**
 * Extract UNSEEN eval lines from the full HF corpus on the VPS.
 *
 * A line counts as unseen iff it does not appear in the training sample
 * (sample_raw.txt). Lines are classified by heuristic (english/code/sql/
 * markdown) and N per domain are written to unseen_eval.json.
 * Benchmark texts are excluded by fingerprint to avoid contamination.
 *
 * Usage: node extract_unseen.mjs <corpus> <sample> <out.json> [perDomain=150]
 */
import { createReadStream } from 'fs';
import { createInterface } from 'readline';
import { writeFileSync } from 'fs';

const [corpus, sample, out] = process.argv.slice(2);
const PER = Number(process.argv[5] ?? process.argv[4] ?? 150);

// Fingerprints of the 8 benchmark texts + boost collocation sentences.
const CONTAM = [
  'quick brown fox', 'rain in spain', 'emergency broadcast', 'brown cow',
  'express()', 'app.get(', 'created_at DESC', 'PRIMARY KEY,name',
  'john@example.com', 'per_page', 'Hello, World!', 'hyprland.conf',
  'git@host:user', 'aicl is Goated',
];

const isCode = (l) =>
  /^\s*(def |function |class |const |let |var |import |from |export |return |if\s*\(|for\s*\(|while\s*\(|elif |print\(|@)\b/.test(l) ||
  /=>|;\s*$|\{\s*$|^\s+\}|^\s*[a-z_][\w.]*\(/.test(l);
const isSql = (l) =>
  /\b(SELECT|INSERT INTO|CREATE TABLE|UPDATE\s+\w+\s+SET|DELETE FROM|ALTER TABLE|GROUP BY|ORDER BY)\b/.test(l);
const isMarkdown = (l) =>
  /^#{1,6} /.test(l) || /^\s*[-*+]\s/.test(l) || /^\|/.test(l) ||
  /\*\*[^*]{2,}\*\*/.test(l) || /^>\s/.test(l) || /^```/.test(l);

function classify(l) {
  if (isSql(l)) return 'sql';
  if (isMarkdown(l)) return 'markdown';
  if (isCode(l)) return 'code';
  return 'english';
}

const usable = (l) => {
  if (l.length < 60 || l.length > 2000) return false;
  let nonAscii = 0;
  for (const c of l) if (c.codePointAt(0) > 127) nonAscii++;
  return nonAscii / l.length < 0.02 && !CONTAM.some((f) => l.includes(f));
};

// Load training-sample lines for exclusion.
console.error('loading sample…');
const seen = new Set();
{
  const rl = createInterface({ input: createReadStream(sample), crlfDelay: Infinity });
  for await (const line of rl) seen.add(line);
}
console.error(`sample lines loaded: ${seen.size}`);

const buckets = { english: [], code: [], sql: [], markdown: [] };
const rl = createInterface({ input: createReadStream(corpus), crlfDelay: Infinity });
let n = 0;
for await (const line of rl) {
  n++;
  if (seen.has(line)) continue;
  if (!usable(line)) continue;
  const d = classify(line);
  if (buckets[d].length < PER) buckets[d].push(line);
}
console.error(`corpus lines: ${n}`);

const counts = {};
for (const [d, lines] of Object.entries(buckets)) {
  counts[d] = lines.length;
  // deterministic shuffle
  for (let i = lines.length - 1; i > 0; i--) {
    const j = (i * 2654435761) % (i + 1);
    [lines[i], lines[j]] = [lines[j], lines[i]];
  }
}
writeFileSync(out, JSON.stringify(buckets, null, 1));
console.log('extracted:', JSON.stringify(counts));

File: scripts/extract_unseen.py
#!/usr/bin/env python3
"""Extract deterministic unseen evaluation lines from a labelled corpus."""

from __future__ import annotations

import argparse
import json
import re
import sys
from pathlib import Path
from typing import TextIO

CONTAMINATION = (
    "quick brown fox",
    "rain in spain",
    "emergency broadcast",
    "brown cow",
    "express()",
    "app.get(",
    "created_at DESC",
    "PRIMARY KEY,name",
    "john@example.com",
    "per_page",
    "Hello, World!",
    "hyprland.conf",
    "git@host:user",
    "aicl is Goated",
)

_CODE_START = re.compile(
    r"(?:def |function |class |const |let |var |import |from |export |return |if\s*\(|for\s*\(|while\s*\(|elif |print\(|@)\b"
)
_CODE_SHAPE = re.compile(r"=>|;\s*$|\{\s*$|^\s+\}|^\s*[a-z_][\w.]*\(")
_SQL = re.compile(
    r"\b(?:SELECT|INSERT INTO|CREATE TABLE|UPDATE\s+\w+\s+SET|DELETE FROM|ALTER TABLE|GROUP BY|ORDER BY)\b"
)
_MARKDOWN = re.compile(r"^#{1,6} |^\s*[-*+]\s|^\||\*\*[^*]{2,}\*\*|^>\s|^```")


def is_code(line: str) -> bool:
    return bool(_CODE_START.search(line) or _CODE_SHAPE.search(line))


def is_sql(line: str) -> bool:
    return bool(_SQL.search(line))


def is_markdown(line: str) -> bool:
    return bool(_MARKDOWN.search(line))


def classify(line: str) -> str:
    if is_sql(line):
        return "sql"
    if is_markdown(line):
        return "markdown"
    if is_code(line):
        return "code"
    return "english"


def usable(line: str) -> bool:
    if len(line) < 60 or len(line) > 2000:
        return False
    non_ascii = sum(ord(char) > 127 for char in line)
    return non_ascii / len(line) < 0.02 and not any(item in line for item in CONTAMINATION)


def _lines(path: Path) -> TextIO:
    return path.open("r", encoding="utf-8", newline="")


def extract(corpus: Path, sample: Path, per_domain: int = 150) -> dict[str, list[str]]:
    if per_domain < 0:
        raise ValueError("per-domain count cannot be negative")
    seen: set[str] = set()
    with _lines(sample) as handle:
        for line in handle:
            seen.add(line.rstrip("\r\n"))
    buckets: dict[str, list[str]] = {"english": [], "code": [], "sql": [], "markdown": []}
    count = 0
    with _lines(corpus) as handle:
        for raw_line in handle:
            count += 1
            line = raw_line.rstrip("\r\n")
            if line in seen or not usable(line):
                continue
            domain = classify(line)
            if len(buckets[domain]) < per_domain:
                buckets[domain].append(line)
    for lines in buckets.values():
        for index in range(len(lines) - 1, 0, -1):
            swap = (index * 2654435761) % (index + 1)
            lines[index], lines[swap] = lines[swap], lines[index]
    print(f"corpus lines: {count}", file=sys.stderr)
    return buckets


def self_check() -> None:
    assert classify("SELECT * FROM users") == "sql"
    assert classify("## heading") == "markdown"
    assert classify("const value = 1;") == "code"
    assert classify("ordinary prose") == "english"
    assert not usable("x" * 59)
    print("extract_unseen self-check: ok")


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("corpus", type=Path, nargs="?")
    parser.add_argument("sample", type=Path, nargs="?")
    parser.add_argument("out", type=Path, nargs="?")
    parser.add_argument("per_domain", type=int, nargs="?", default=150)
    parser.add_argument("--self-check", action="store_true")
    args = parser.parse_args(argv)
    if args.self_check:
        self_check()
        return 0
    if args.corpus is None or args.sample is None or args.out is None:
        parser.error("corpus, sample, and out are required unless --self-check is used")
    buckets = extract(args.corpus, args.sample, args.per_domain)
    args.out.parent.mkdir(parents=True, exist_ok=True)
    args.out.write_text(json.dumps(buckets, ensure_ascii=False, indent=1) + "\n", encoding="utf-8")
    print("extracted:", json.dumps({key: len(value) for key, value in buckets.items()}))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
